File: pt5/1120.maximum-average-subtree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maximumAverageSubtree(root *TreeNode) float64 {
	best := 0.0

	var dfs func(*TreeNode) (int, int)
	dfs = func(node *TreeNode) (int, int) {
		if node == nil {
			return 0, 0
		}

		leftSum, leftCount := dfs(node.Left)
		rightSum, rightCount := dfs(node.Right)
		totalSum := leftSum + rightSum + node.Val
		totalCount := leftCount + rightCount + 1
		average := float64(totalSum) / float64(totalCount)
		if average > best {
			best = average
		}
		return totalSum, totalCount
	}

	dfs(root)
	return best
}

File: pt5/1120.maximum-average-subtree.py
from __future__ import annotations

from typing import Optional, Tuple


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maximumAverageSubtree(self, root: Optional[TreeNode]) -> float:
        best = 0.0

        def dfs(node: Optional[TreeNode]) -> Tuple[int, int]:
            nonlocal best
            if node is None:
                return 0, 0

            left_sum, left_count = dfs(node.left)
            right_sum, right_count = dfs(node.right)
            total_sum = left_sum + right_sum + node.val
            total_count = left_count + right_count + 1
            best = max(best, total_sum / total_count)
            return total_sum, total_count

        dfs(root)
        return best

